designs/0005-static-matrices.md

```diff
@@ -0,0 +1,240 @@
+implementation. As an incomplete substitute, a list of functions with reverse mode
+implementations that are being converted to work with varmat is here:
+https://github.com/stan-dev/math/issues/2101 . A pull request
+([here](https://github.com/stan-dev/math/pull/2214)) is up that adds support for
+varmat to all the non-multivariate distribution functions.
+
+## Assigning a varmat
+
+Something in the original design doc that is not necessary anymore is the limitation
+that subsets of varmat variables cannot be assigned efficiently.
+
+The efficiency loss was so great in the original design (requiring a copy of the entire
+varmat) that this idea was discarded completely.
+
+There is a strategy available now to assign to subsets of varmat variables that is
+more efficient than this.
+
+The idea is that when part of a varmat is assigned to, the values in that part of the
+varmat are saved into the memory arena before being overwritten, and a call is pushed
+onto the chaining stack that will restore the values back from the arena and set the
+respective adjoints of the varmat entries to zero.
+
+The assumptions that makes this work is after a subset of a varmat is overwritten,
+that part of the varmat cannot be used in any further calculations, and so the adjoints
+of that part of the matrix will be zero in the reverse pass at the time of assignment.
+
+This means that assignment into a varmat would invalidate any views into that
+varmat's memory.
+
+# Stanc3 implementation
+[stanc3-implementation]: #stanc3-implementation
+
+The language implementation of varmat is yet undefined. It was left unresolved because
+enough was clear from the initial discussions of how to handle the Math implementation
+regardless of how the language was handled.
+
+At this point, the Math maturing to the point that it is possible to compile a model using
+varmats values with a development version of the Stanc3 compiler (available
+[here](https://github.com/stan-dev/stanc3/pull/755), though it may take some effort to run since
+it is a work in progress). This Stanc3 compiler defines any variable with a name ending in
+`_varmat` to use a varmat type.
+
+A basic bernoulli MRP model with five population level covariates and seven group terms is available
+[here](https://github.com/bbbales2/computations_in_glms/blob/master/benchmark_model/mrp_varmat.stan).
+
+The initial varmat implementation of this model achieves a per-gradient speedup of around two.
+
+Now that varmat variables can be assigned, it is possible that the compiler silently replace every
+matvar in Stan with a varmat, and none of the standard Stan use cases would be affected.
+
+There are four major implementation strategies available at the Stanc3 level.
+
+1. Implement a new Stan type to differentiate explicitly between matvar and varmat types
+
+2. Use matvar types whenever there is any function in the Stan program that does not support varmat
```

Comment:
Once all the functions support varmat this is the same as 4.

Reply:
I don't think we'll ever reach the 100% conversion point.

Reply:
I added a note that 2, 3, and 4 all become the same thing if varmat is supported everywhere.